pallets/file-system/src/utils.rs

```diff
@@ -176,10 +203,104 @@ where
         Ok(())
     }
 
+    pub(crate) fn do_bsp_confirm_storing(
+        who: T::AccountId,
+        location: FileLocation<T>,
+        root: FileKey<T>,
+        proof: Proof<T>,
+    ) -> DispatchResult {
+        let bsp =
+            <T::Providers as storage_hub_traits::ReadProvidersInterface>::get_provider(who.clone())
+                .ok_or(Error::<T>::NotAProvider)?;
+
+        // Check that the provider is indeed a BSP.
+        ensure!(
+            <T::Providers as storage_hub_traits::ReadProvidersInterface>::is_bsp(&bsp),
+            Error::<T>::NotABsp
+        );
+
+        // Check that the storage request exists.
+        let file_metadata =
+            <StorageRequests<T>>::get(&location).ok_or(Error::<T>::StorageRequestNotFound)?;
+
+        expect_or_err!(
+                    file_metadata.bsps_confirmed < file_metadata.bsps_required,
+                    "Storage request should never have confirmed bsps equal to or greater than required bsps, since they are deleted when it is reached.",
+                    Error::<T>::StorageRequestBspsRequiredFulfilled,
+                    bool
+                );
+
+        // Check that the BSP has volunteered for the storage request.
+        ensure!(
+            <StorageRequestBsps<T>>::contains_key(&location, &who),
+            Error::<T>::BspNotVolunteered
+        );
+
+        // Check that the storage provider has not already confirmed storing the file.
+        ensure!(
+            !<StorageRequestBsps<T>>::get(&location, &who)
+                .expect("BSP should exist since we checked it above")
+                .confirmed,
+            Error::<T>::BspAlreadyConfirmed
+        );
+
+        let mut file_metadata =
+            <StorageRequests<T>>::get(&location).ok_or(Error::<T>::StorageRequestNotFound)?;
+
+        // Check that the number of confirmed bsps is less than the required bsps and increment it.
```

Comment:
This is probably an old comment, as you're not actually checking that the confirmed BSPs is less than the required. In fact, it is correct that you're not checking it, because it shouldn't be possible for the confirmed BSPs to be greater or equal than the required. In that case, there shouldn't even be a storage request anymore.

Reply:
I support that you check it anyway, but with `expect_or_err`, as it is not an expected state of the chain.

Reply:
What you are checking is an overflow in the `checked_add`, so the error should be an overflow, not `StorageRequestBspsRequiredFulfilled`.

Reply:
Oh! Thanks for spotting this. 🙏🏼